src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -1232,22 +1235,54 @@ class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
     filterVariants(p)
   }
 
-  def filterVariantsList(input: String, keep: Boolean): VariantSampleMatrix[T] = {
-    copy(
-      rdd = rdd
-        .orderedLeftJoinDistinct(Variant.variantUnitRdd(sparkContext, input).toOrderedRDD)
-        .mapPartitions({ it =>
-          it.flatMap { case (v, ((va, gs), o)) =>
-            o match {
-              case Some(_) =>
-                if (keep) Some((v, (va, gs))) else None
-              case None =>
-                if (keep) None else Some((v, (va, gs)))
-            }
-          }
-        }, preservesPartitioning = true)
-        .asOrderedRDD
-    )
+  def filterVariantsList(variants: java.util.ArrayList[Variant], keep: Boolean): VariantSampleMatrix[T] =
+    filterVariantsList(variants.asScala.toSet, keep)
+
+  def filterVariantsList(variants: Set[Variant], keep: Boolean): VariantSampleMatrix[T] = {
+    val partitionVariants = variants
+      .groupBy(v => rdd.orderedPartitioner.getPartition(v))
+      .toArray
+      .sortBy(_._1)
+
+    val adjRDD = new AdjustedPartitionsRDD[RowT](rdd,
+      partitionVariants.map { case (oldPart, variantsSet) =>
+        Array(Adjustment[RowT](oldPart,
+          _.filter { case (v, _) =>
+            Filter.keepThis(variantsSet.contains(v), keep)
+          }))
+      })
+
+    val adjRangeBounds =
+      if (partitionVariants.isEmpty)
+        Array.empty[Locus]
+      else
+        partitionVariants.init.map { case (oldPart, _) =>
+          rdd.orderedPartitioner.rangeBounds(oldPart)
+        }
+
+    copy(rdd = OrderedRDD(adjRDD, OrderedPartitioner(adjRangeBounds, partitionVariants.length)(
+      rdd.kOk)))
+  }
+
+  def filterVariantsKT(kt: KeyTable, keep: Boolean): VariantSampleMatrix[T] = {
+    if (kt.keyFields.length != 1 || kt.keyFields(0).typ != TVariant)
```

Comment:
So, if instead of using leftOrderedJoinDistinct and futzing the Option business, if we do the add/remove with linear scans on the iterators, I don't think this buys you much.  When it is small, the shuffle on the RHS is no worse than the collect, and you can short circuit the iterator scan when the RHS is empty (or ends early).  I think we should benchmark this.